go/database/mpt/proof.go

```diff
@@ -0,0 +1,197 @@
+}
+
+// GetAccountProof returns the account proof.
+func (p *StorageProofs) GetAccountProof() AccountProof {
+	return p.accountProof
+}
+
+// GetStorageProof returns the storage proof for the input storage key.
+func (p *StorageProofs) GetStorageProof(key common.Key) StorageProof {
+	return p.storage[key]
+}
+
+// CreateAccountProof creates a witness proof for the account at the given address.
+// It iterates the MPT from the root down to the leaf node following the address.
+// Each node encountered on the path is hashed and stored in a proof.
+func (p *ProofSource) CreateAccountProof(address common.Address) (AccountProof, error) {
+	visitor := &nodeVisitor{
+		hasher:     p.hasher,
+		nodeSource: p.nodeSource,
+	}
+	exists, err := VisitPathToAccount(p.nodeSource, p.root, address, visitor)
+	if err != nil {
+		return AccountProof{}, err
+	}
+	if !exists {
+		return AccountProof{}, visitor.err
+	}
+
+	return AccountProof(visitor.proof), visitor.err
+}
+
+// CreateStorageProofs creates a witness proof for the input account
+// and possibly many storage slots of the same account under the input storage keys.
+// After getting an account proof, it iterates the MPT from the account down to the
+// leaf node following all input keys.
+// Each node encountered on the path is hashed and stored in a proof.
+// The object representing the account and storage proofs is returned.
+func (p *ProofSource) CreateStorageProofs(address common.Address, keys ...common.Key) (StorageProofs, error) {
+	accountVisitor := &nodeVisitor{
+		hasher:     p.hasher,
+		nodeSource: p.nodeSource,
+	}
+
+	var storageVisitError error
+
+	proofs := StorageProofs{storage: make(map[common.Key]StorageProof, len(keys))}
+	exists, err := VisitPathToAccount(p.nodeSource, p.root, address, MakeVisitor(func(node Node, info NodeInfo) VisitResponse {
+		if res := accountVisitor.Visit(node, info); res == VisitResponseAbort {
+			return VisitResponseAbort
+		}
+		// if reached account, prove storage keys and terminate.
+		switch account := node.(type) {
+		case *AccountNode:
+			for _, key := range keys {
+				storageVisitor := &nodeVisitor{
+					hasher:     p.hasher,
+					nodeSource: p.nodeSource,
+				}
+				exists, err := VisitPathToStorage(p.nodeSource, &account.storage, key, storageVisitor)
+				if err != nil || storageVisitor.err != nil {
+					storageVisitError = errors.Join(storageVisitError, storageVisitor.err, err)
+					return VisitResponseAbort
+				}
+				if exists {
+					proofs.storage[key] = StorageProof(storageVisitor.proof)
+				}
+			}
+
+			// all requested keys done, terminate.
+			proofs.accountProof = AccountProof(accountVisitor.proof)
+			return VisitResponseAbort
+		}
+
+		return VisitResponseContinue
+	}))
+
+	if err != nil || storageVisitError != nil || accountVisitor.err != nil {
+		return StorageProofs{}, errors.Join(storageVisitError, accountVisitor.err, err)
+	}
+
+	if !exists {
+		return StorageProofs{storage: make(map[common.Key]StorageProof)}, nil
```

Comment:
Is creating the map for an empty proof necessary?

Reply:
it is for a more user friendly interface. If the proof is empty and user calls:  `proof.GetStorageProof(key)` and the map was not initialised, it would panic. This way it just returns an empty proof.

Reply:
hm, actually I checked that it does not panic when the map is not initialised. Is it a feature of go that getting a key from a map, which is nil does not fail?

Reply:
I believe nil map does not panic unless you assign yo it.

Reply:
Right. Check this test: 
```
func TestEmptyMap_Panic(t *testing.T) {
	var m map[string]string
	// no panic
	_ = m["a"]

	// panics
	m["a"] = "b"
}
```

It fails with `panic: assignment to entry in nil map`

Reply:
Yes, reading from `nil` maps is perfectly fine and should be utilized, as in Go default values should be valid states wherever possible.

Reply:
ok, removed